From: Avi Kiviti <avi@qumranet.com>
Date: Tue, 15 Jan 2008 15:02:22 +0200
Subject: kvm: check need_resched() inside the irq disabled region

The missing need_resched() check inside the irq_disabled region can cause
long latencies, if a interrupt with reschedule request happens between 
preempt_disable() and the local_irq_disable(). In fact the interrupts are
disabled inside prepare_guest_switch(), so the race window is rather small.

This can be further optimized, but it fixes the bug for now.

Mainline bug is fixed in a similar way.

Signed-off-by: Avi Kiviti <avi@qumranet.com>
Signed-off-by: Thomas Gleixner <tgxl@linutronix.de>

---
 drivers/kvm/kvm_main.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.24-rt1/drivers/kvm/kvm_main.c
===================================================================
--- linux-2.6.24-rt1.orig/drivers/kvm/kvm_main.c	2008-01-25 15:06:47.000000000 -0500
+++ linux-2.6.24-rt1/drivers/kvm/kvm_main.c	2008-01-25 15:07:06.000000000 -0500
@@ -2010,6 +2010,13 @@ again:
 
 	local_irq_disable();
 
+	if (need_resched()) {
+		local_irq_enable();
+		preempt_enable();
+		r = 1;
+		goto out;
+	}
+
 	if (signal_pending(current)) {
 		local_irq_enable();
 		preempt_enable();
